Rubberduck.CodeAnalysis/Inspections/Concrete/AssignmentNotUsedInspection.cs

```diff
@@ -0,0 +1,114 @@
+using System.Collections.Generic;
+using Rubberduck.Inspections.Abstract;
+using Rubberduck.Parsing.Inspections.Abstract;
+using Rubberduck.Parsing.VBA;
+using Rubberduck.Inspections.CodePathAnalysis;
+using Rubberduck.Parsing.Symbols;
+using Rubberduck.Inspections.CodePathAnalysis.Nodes;
+using System.Linq;
+using Rubberduck.Inspections.Results;
+
+namespace Rubberduck.Inspections.Concrete
+{
+    public sealed class AssignmentNotUsedInspection : InspectionBase
+    {
+        private readonly Walker _walker;
+
+        public AssignmentNotUsedInspection(RubberduckParserState state, Walker walker)
+            : base(state) {
+            _walker = walker;
+        }
+
+        protected override IEnumerable<IInspectionResult> DoGetInspectionResults()
+        {
+            var variables = State.DeclarationFinder.UserDeclarations(DeclarationType.Variable);
+
+            var nodes = new List<IdentifierReference>();
+            foreach (var variable in variables)
+            {
+                var tree = _walker.GenerateTree(variable.ParentScopeDeclaration.Context, variable);
+
+                nodes.AddRange(GetReferences(tree, variable));
+            }
+
+            return nodes
+                .Select(issue => new IdentifierReferenceInspectionResult(this, Description, State, issue))
+                .ToList();
+        }
+
+        private List<IdentifierReference> GetReferences(INode node, Declaration declaration)
+        {
+            var nodes = new List<IdentifierReference>();
+
+            var blockNodes = GetBlocks(node);
+            foreach (var block in blockNodes)
+            {
+                INode lastNode = default;
+                foreach (var flattenedNode in GetFlattenedNodes(block))
+                {
+                    if (flattenedNode is AssignmentNode &&
+                        lastNode is AssignmentNode)
+                    {
+                        nodes.Add(lastNode.Reference);
+                    }
+
+                    lastNode = flattenedNode;
+                }
+
+                if (lastNode is AssignmentNode &&
+                    GetFirstNonGenericNode(block.Children[0]) is DeclarationNode)
+                {
+                    nodes.Add(lastNode.Reference);
+                }
+            }
+
+            return nodes;
+        }
+
+        private IEnumerable<INode> GetFlattenedNodes(INode node)
```

Comment:
this might be best off as an extension method on `INode`. You're basically trying to expose an IEnumerable over the flattened tree, which shouldn't need to be implemented by the Inspections traversing the tree...

Reply:
Yep, thought about that.